d2/src/main/java/com/linkedin/d2/balancer/subsetting/ZKDeterministicSubsettingMetadataProvider.java

```diff
@@ -0,0 +1,113 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.subsetting;
+
+import com.linkedin.common.callback.FutureCallback;
+import com.linkedin.d2.balancer.LoadBalancerState;
+import com.linkedin.d2.balancer.LoadBalancerStateItem;
+import com.linkedin.d2.balancer.properties.UriProperties;
+import java.net.URI;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import org.apache.http.annotation.GuardedBy;
+
+
+/**
+ * Listens to the peer cluster in ZooKeeper and provides deterministic subsetting strategy with
+ * the metadata needed
+ */
+public class ZKDeterministicSubsettingMetadataProvider implements DeterministicSubsettingMetadataProvider
+{
+  private final String _clusterName;
+  private final URI _nodeUri;
+  private final LoadBalancerState _loadBalancerState;
+  private final long _timeout;
+  private final TimeUnit _unit;
+
+  private final Object _lock = new Object();
+
+  @GuardedBy("_lock")
+  private long _clusterGenerationId = -1;
+  private DeterministicSubsettingMetadata _subsettingMetadata;
+
+  public ZKDeterministicSubsettingMetadataProvider(String clusterName,
+                                      URI nodeUri,
+                                      LoadBalancerState loadBalancerState,
+                                      long timeout,
+                                      TimeUnit unit)
```

Comment:
I see the problem. There's a circular dependency between ZKDeterministicSubsettingMetadataProvider and LoadBalancerState. I resolved this circular dependency by removing LoadBalancerState from the constructor of ZKDeterministicSubsettingMetadataProvider, and passing it as an argument to ZKDeterministicSubsettingMetadataProvider.getSubsettingMetadata(state). Then have LoadBalancerState depend on SubsettingMetadataProvider directly.

Then in container, with clusterName, nodeUri, timeout and unit, we're able to instantiate ZKDeterministicSubsettingMetadataProvider and pass it to d2 config. D2 config will use the ZKDeterministicSubsettingMetadataProvider to instantiate LoadBalancerState. Then in LoadBalancerState, it will create an instance of SubsettingStrategyFactory using ZKDeterministicSubsettingMetadataProvider and itself.